plugins/demoplugin.c

```diff
@@ -35,29 +35,19 @@
 #include "geanyplugin.h"	/* plugin API, always comes first */
 #include "Scintilla.h"	/* for the SCNotification struct */
 
-
-/* These items are set by Geany before plugin_init() is called. */
-GeanyPlugin		*geany_plugin;
-GeanyData		*geany_data;
-
-
-/* Check that the running Geany supports the plugin API version used below, and check
- * for binary compatibility. */
-PLUGIN_VERSION_CHECK(147)
-
-/* All plugins must set name, description, version and author. */
-PLUGIN_SET_INFO(_("Demo"), _("Example plugin."), "0.1" , _("The Geany developer team"))
-
-
 static GtkWidget *main_menu_item = NULL;
 /* text to be shown in the plugin dialog */
 static gchar *welcome_text = NULL;
 
 
-
 static gboolean on_editor_notify(GObject *object, GeanyEditor *editor,
 								 SCNotification *nt, gpointer data)
 {
+	/* Geany passes the plugin's GeanyPlugin pointer as user_data if it did not set
+	 * it explicitely. This is specific to PluginCallbacks. */
```

Comment:
> tbh, I don't see a lot of plugins (outside proxy plugins) making use set_data(). _Especially_ for large and complex plugins just using globals is convenient and effective.

On the contrary I myself would try to use this alone and get rid of as much global state as reasonably possible.  I didn't try on very large plugins of course, but I believe it's totally feasible and IMO better (things have to be better separated, but that's hardly ever a bad thing).

Also imagine in the future we want multi-window single-instance Geany, it could be a lot less effort to have global-state-free plugins so we could start activating them once for each window :)  Yes I'm dreaming.

> I don't mind changing this though. What to do when no data was set?

I'd say do the exact same as the data arg for the init/cleanup etc: just pass the data, so default to NULL.  Defaulting to something else if not set is likely to be more confusing than useful.